src/main/java/com/zero_x_baadf00d/play/module/redis/cache/AsyncCacheRedisImpl.java

```diff
@@ -0,0 +1,161 @@
+ * @see PlayRedis
+ * @since 20.11.05
+ */
+@Singleton
+public class AsyncCacheRedisImpl implements AsyncCacheApi {
+
+    /**
+     * Default prefix mapping content saved in Redis.
+     *
+     * @since 20.11.05
+     */
+    private static final String DEFAULT_PREFIX = "play.cache.";
+
+    /**
+     * The {@link PlayRedis} injected implementation.
+     *
+     * @since 20.11.05
+     */
+    private final SyncCacheRedisImpl syncCacheApi;
+
+    @Inject
+    public AsyncCacheRedisImpl(final SyncCacheRedisImpl syncCacheApi) {
+        this.syncCacheApi = syncCacheApi;
+    }
+
+    /**
+     * Get the {@link PlayRedis} implementation class.
+     *
+     * @return A {@link PlayRedis} implementation.
+     * @since 20.11.05
+     */
+    public PlayRedis asPlayRedis() {
+        return this.syncCacheApi.asPlayRedis();
+    }
+
+    @Override
+    public <T> CompletionStage<Optional<T>> get(final String key) {
+        return supplyAsync(() -> this.sync().get(this.prepareKey(key)));
+    }
+
+    @Override
+    public <T> CompletionStage<T> getOrElseUpdate(final String key, final Callable<CompletionStage<T>> callable,
+                                                  final int expiration) {
+        return CompletableFuture.supplyAsync(() -> this.sync().<T>get(this.prepareKey(key)))
+            .thenCompose(optional -> {
+                if (optional.isPresent())
+                    return completedFuture(optional.get());
+
+                try {
+                    return callable.call()
+                        .thenCompose(t ->
+                            this.set(key, t).thenApply(done -> t));
+                } catch (Exception e) {
+                    e.printStackTrace();
+                    throw new CompletionException(e);
+                }
+            });
+    }
+
+    @Override
+    public <T> CompletionStage<T> getOrElseUpdate(final String key, final Callable<CompletionStage<T>> callable) {
+        return this.getOrElseUpdate(key, callable, 0);
+    }
+
+    @Override
+    public CompletionStage<Done> set(final String key, final Object o, final int expiration) {
+        return supplyAsync(() -> {
+            this.sync().set(this.prepareKey(key), o, expiration);
+            return Done.done();
+        });
+    }
+
+    @Override
+    public CompletionStage<Done> set(final String key, final Object o) {
+        return this.set(key, o, 0);
+    }
+
+    @Override
+    public CompletionStage<Done> remove(final String key) {
+        return supplyAsync(() -> {
+            this.sync().remove(this.prepareKey(key));
+            return Done.done();
+        });
+    }
+
+    @Override
+    public CompletionStage<Done> removeAll() {
+        return supplyAsync(() -> {
+            try (final Jedis connection = this.asPlayRedis().getConnection(0)) {
+                // Search all keys in Redis.
+                // https://redis.io/commands/keys
+                connection.keys(DEFAULT_PREFIX + "??").forEach(this::remove);
```

Comment:
I think you have to replace `??` with `*`. According to the documentation `??` will only match a two characters word.

By example, if you have the following keys :
- play.cache.p1
- play.cache.cn
- play.cache.index
- play.cache.dashboard

The pattern `DEFAULT_PREFIX + "??"` will only match "play.cache.p1" and "play.cache.cn". The pattern  `DEFAULT_PREFIX + "*"` will match all keys beginning with DEFAULT_PREFIX

Reply:
You're right!